app/client/packages/design-system/widgets/src/components/ToolbarButtons/src/ToolbarButtons.tsx

```diff
@@ -37,11 +37,14 @@ const _ToolbarButtonsInner = <T extends ToolbarButtonsItem>(
     domRef,
   );
 
-  let children = [...state.collection];
-  const menuChildren = (props.items as ToolbarButtonsItem[]).slice(
-    visibleItems,
+  const menuChildren = useMemo(
+    () => (props.items as ToolbarButtonsItem[]).slice(visibleItems),
+    [props.items, visibleItems],
   );
-  children = children.slice(0, visibleItems);
+  const children = useMemo(() => {
+    const children = [...state.collection];
+    return children.slice(0, visibleItems);
```

Comment:
Do we need a transient variable? 
```suggestion
    return [...state.collection].slice(0, visibleItems);
```

Even if we do lets rename it to make it more readable - we have `children` in both outer and inner scope.

Reply:
> Do we need a transient variable?

Because collection is not iterable we have to spread it. Anyway we can do something like `[...state.collection]. slice`